Adds minor roundstart traits! (ala CDDA, etc.)

Xhuis opened the pull request:
:cl: Xhuis
add: Traits! You can now select up to three positive, negative, and neutral traits in the character setup. You can choose up to six traits based on combinations varying by the amount of points you earn with negative traits and spend with positive ones.
/:cl:

![](https://i.imgur.com/clMQu2R.png)

This is something I've wanted to have for a long time, and with the trait system now implemented, it seemed like the perfect time to implement it by building off of that framework.

The goal is to provide more character customization options that change the way you play in minor (or, with some negatives, major) ways. Someone who can't taste any food might not worry too much over what they eat, and someone who sees more clearly in the dark might be more comfortable without a light. They aren't meant to provide a tangible benefit that would be game-changing or incredibly useful; just to provide some quirks to your character to make the game more interesting.

Currently, each trait has a point value. You start with a balance of 0, and so can't purchase any traits. Points are gained by taking negative traits, and spent by taking positive ones with the points gained from negative traits. You can only have a maximum of six traits at a time.

Once the round starts, you can't change your traits (unless you're in the lobby) until the next round. This is to prevent you from swapping traits between being a corpse and being cloned.

Current positive traits:
* **Alcohol Tolerance (-1):** You get drunk more slowly and suffer less negative effects (i.e. no jittering from thirteen loko.)
* **Freerunning (-2):** You vault tables a little bit faster.
* **Light Step (-1):** You take less damage from stepping on sharp things, and shards make less noise when you walk over them.
* **Night Vision (-1):** Lets you see slightly clearer in darkness. *The default sight range is still 3x3, even though the icon diff bot will show the picture as very large.*
* **Self-Aware (-2):** Inspecting yourself for injuries shows brute and burn in their exact amounts, and vague descriptions of toxin and oxygen damage.
* **Skittish (-2):** You can flee at the first sight of danger. Ctrl-shift-clicking a locker or crate, open or closed, will cause you to instantly dive into it and close the door behind you!
* **Spiritual (-1):** Gives you a different name next to your prayer like cultists and chaplains get.
* **Voracious (-1):** You can eat and drink twice as fast.

Current negative traits:
* **Arachnophobia (+1):** Gives you a fear of spiders using the phobia system.
* **Heavy Sleeper (+1):** Unconsciousness and sleeping effects last longer for you.
* **Nearsighted (+1):** You spawn with incurable, permanent nearsightedness! You also spawn with a pair of glasses, so you can be the QT nerd girl you always wanted to be.
* **Pacifist (+2):** This applies the pacifism trauma, permanently. You are unable to harm anyone. This *does* cleanse, however, if you're an antagonist.
* **Poor Aim (+1):** You can't aim well with guns and all shots you fire will be slightly off-target.
* **Prosopagnosia (+1):** You can't recognize faces, and examining anyone will display their name as `Unknown` regardless of what they're wearing.
* **Prosthetic Limb (+1):** An accident caused you to lose a limb! One of your legs or arms (chosen randomly) will be a surplus prosthetic version.
* **Reality Dissociation Syndrome (+2):** This is an interesting one that will largely change your rounds. Every tick, there's a 1% chance you will suffer some sort of "dissocation" with reality. Most often this will be normal hallucinations, but effects can range from intense hallucinations (ala nar-sie rising) to seeing ghosts for a flicker of time to fourth-wall breaking threats in your chat box. Mindbreaker toxin will suppress these effects; additionally, having this trait nullifies the hallucination effects of mindbreaker.
  * This is not a license to grief via roleplayed insanity and there is a chat disclaimer about such.
* **Social Anxiety (+1):** Talking to people is difficult for you, and you can often stammer and occasionally lock up and become incapable of talking at all.

Current neutral traits:
* **Ageusia:** You can't taste any food or reagents. Toxic things will still poison you.
* **Deviant Tastes:** Inverts your liked and disliked foods.

There's a lot of traits I could add into this, but I wanted to ensure that the system itself was accepted before I made a very large list.

Traits can be enabled or disabled with the `ROUNDSTART_TRAITS` option in the game options txt.

Commit:
Roundstart traits

Ordonis commented:
God damn, this will be horribly abused and I cannot wait

optimumtact commented:
Deviant tastes enabling liking pineapple on pizza when?

Hatterhat commented:
i want to bully qt nerd grills, speedmerge

ike709 commented:
Not a fan of the menu formatting.

Xhuis commented:
@ike709 What would you suggest?

ike709 commented:
Move the gain/lose buttons to the left side, align them, and align the names. Or left-align the names, right-align the gain/lose, and center-align the descriptions. As long as the names are vertically aligned with eachother, and the buttons are vertically aligned with eachother.

Change the current traits to be colored based on if they're positive/negative/neutral for readability. Not sure how feasible this is.

Add a button to clear current traits, and a random traits button.

That's just my two cents.

TechnoAlchemisto commented:
Traits should probably not all be worth 1 point. Really debilitating traits should probably let you take more than one positive traits and visa versa. 

Doing that would be a bitch to balance, but would actually give some reason to take certain traits besides gimmicks.. Otherwise it's just going to be night vision + heavy sleeper.

ike709 commented:
Also, neutral should be between positive and negative.

Midaychi commented:
Any chance of getting a way to see as a ghost if someone has pacifist or RDS?  I feel like ghosts not knowing this would cause more problems then them knowing.

MMMiracles commented:
pacifist nuke op when

Xhuis commented:
@Midaychi That's a good idea, I'll put that in the next commit.
@TechnoAlchemisto I'd like to add some sort of point system, but I wanted to see how people responded to the concept first.
@ike709 I'll try some reformatting and post back.

arsserpentarium commented:
Needs Quistinnus moodlets now.
good PR.

Bawhoppen commented:
I expected them to be *more* minor than even this. Like the "Be Italian" or "Like Pineapple Pizza" and other stuff. I guess we can do that too.

Also, the not tasting food is definitely very much so a positive trait that I can see people using the powergame with. Not getting sick when eating something that you don't like, is pretty handy.

Xhuis commented:
I figured it was moreso a neutral because it also nullifies the effects of positive foods as well; a benefit with a drawback, both self-contained. It also has the negative of not allowing you to taste reagents, so you don't know what you're drinking from a beaker.

Repukan commented:
Opens a floodgate to unknown edges with no clear sign of being edges which are bad design in and of themselves.

Will be fun to see this being abused to hell and back and seeing everyone take the same combos as CDDA

Bawhoppen commented:
I mean, it's not huge either way, but I wouldn't surprised if it's not used for a benefit more than a hindrance

Xhuis commented:
Why do you keep giving design insights into a game you're no longer allowed to play?

outbools commented:
because he still want to look smart

Repukan commented:
Brilliant response, I can see you're getting better at deflecting and not responding criticism

outbools commented:
It would be great if you can see traits in Medical Records

Xhuis commented:
If you want me to respond to your criticism, then I'd like to say that I respectfully disagree with your comment about it and I don't think that this will cause major issues because I'm balancing them around specifically not doing so.
@outbools I could do that, but I'd need to find a more efficient way to track the traits a mob has first.

outbools commented:
@Repukan >Opens a floodgate to unknown edges with no clear sign of being edges which are bad design in and of themselves.

That's not a criticism, you're just whining about new things. Yes, new things can be abused and yes, even you don't know HOW they will be abused. Cease this speculation.

Repukan commented:
>I'm balancing them around specifically not doing so.

Just looking at this heavy sleeper seems like a no brainer since stuns and sleeping are basically the end of the combat for anyone.

It'll either be bloat for stuff that really nobody except for a couple people will bother with since it makes no difference, or another hellish slug of meta traits to take and the player base suffering through another clockcult tier period of brokenness. 

The real worry here is this stuff won't be contained to its own gamemode.

Repukan commented:
My opinion, and I think this can resonate with some people, is that people aren't enjoying the shit show that is clock cult, and having you undertake another system to produce and mantain, with how stellar the last one was, is a risky proposition, with the players fun at stake.

Repukan commented:
You're free to dismiss this since I'm permabanned. Or maybe ask someone in sligcord to write a rebuttal for you, that'll work better, no?

outbools commented:
Also, instead of Arachnophobia - better use already existing Trauma datum(#35436), right? I mean, better make a Random Phobia trait then just a fear of spiders?

Iamgoofball commented:
@Repukan sligcord is retarded but a feature like this will be fine especially if we make the negative traits heavy shit

improvedname commented:
Just make it so.
The more negative traits you choice (will make your gameplay harder) the more positive you can pick with just the maximum of 5 trait slots

improvedname commented:
Also add insect phobia so you be scared of moths and fly's disgusting creatures

kevinz000 commented on code/datums/traits/_trait.dm:
```diff
@@ -0,0 +1,82 @@
+Commented version of Nearsighted to help you add your own traits
+Use this as a guideline
+
+/datum/trait/nearsighted
```
neersighted

ExcessiveUseOfCobblestone replied:
is this a joke

Xhuis replied:
is this loss?

optimumtact replied:
this should be called trait/errorage in honour of the time he tried to make all engineers shortsighted lol

Xhuis replied:
It was only a 25% chance, thank you very MUCH.

PKPenguin321 commented:
>Voracious
>vore

😠

Xhuis commented:
I was thinking about renaming it to Gourmand.

XDTM commented on code/datums/traits/negative.dm:
```diff
@@ -0,0 +1,110 @@
+	if(prob(1)) //we'll all be mad soon enough
+		madness()
+
+/datum/trait/insanity/proc/madness(mad_fools)
```
Cmon we have a full hallucination system

Xhuis replied:
Do you know where the files for it are? I wanted to put them in that way but I couldn't find them after some searching

PKPenguin321 commented:
Serious feedback, the UI could be better. I noticed right away that it's hard to tell at a glance what abilities you've currently got selected.

XDTM commented on code/datums/traits/negative.dm:
```diff
@@ -0,0 +1,110 @@
+	name = "Pacifist"
+	desc = "The thought of violence makes you sick. So much so, in fact, that you can't hurt anyone."
+	value = -1
+	mob_trait = TRAIT_PACIFISM
```
Use the trauma, it does the same but it's interactable in some way

XDTM commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -587,7 +595,7 @@ All effects don't start immediately, but rather get worse over time; the rate is
 	if(ishuman(M))
 		var/mob/living/carbon/human/H = M
-		if(H.dna.check_mutation(DWARFISM))
+		if(H.dna.check_mutation(DWARFISM) || H.has_trait(TRAIT_ALCOHOL_TOLERANCE))
```
Might want to make dwarfism give the trait since it's there now

XDTM commented on code/datums/traits/neutral.dm:
```diff
@@ -0,0 +1,28 @@
+	name = "Ageusia"
+	desc = "You can't taste anything! Toxic food will still poison you."
+	value = 0
+	mob_trait = TRAIT_AGEUSIA
```
Not in scope, but this could be a nice minor brain trauma

coiax replied:
That is something that can happen to people when they have strokes.

AnturK commented on code/datums/traits/good.dm:
```diff
@@ -0,0 +1,58 @@
+//predominantly positive traits
+//this file is named weirdly so that positive traits are listed above negative ones
```
No.

Xhuis replied:
What part is no? I named it `good.dm` so it alphabetized over `negative.dm`.

AnturK replied:
Anything relying on compile time for in-game ordering shouldn't exist. Unless you just mean in some editor view.

Xhuis replied:
It's for the list to go
`positive`
`negative`
`neutral`
It has no intentional impact on the way things look in the actual panel.

AnturK replied:
You're still relying on compilation order here, use some explicit priority. (Sorting by value being the obvious one)

RemieRichards commented on code/datums/traits/_trait.dm:
```diff
@@ -42,6 +41,13 @@
+	for(var/V in roundstart_traits)
+		var/datum/trait/T = V
+		dat += T.name
+	return "[dat.Join(", ")]" //not sure if I need the quotes here but best to play it safe
```
you dont, ```return dat.Join(", ")```

Anonmare commented:
Could add a permanent phobia as a negative trait. Antagonists should probably be cleared of it, so revs don't break down at the sight of the heads of staff or cultists being terrified every time they use their magic

XDTM commented:
Make all traumas available! Can't wait for a paralytic captain that has to be wheeled away from ops by the HoS

Mey-Ha-Zah commented:
You should give Ashwalkers Light Step. It would help them a little so cactus don't end their life like they often do.

darkpaladin109 commented:
Aww, you can only have up to 6 traits. My dreams of having a character with all the negative traits turned on is ruined..

BeeSting12 commented:
Add psychopath where you ICly dont care about killing people so you can validly valid people

coiax commented:
Can we have Nicotine Addict and Alcohol Addict? I admit our addiction systems aren't really in a place for light addictions.

But it would be nice for the Barkeeper.

CosmicScientist commented:
@Anonmare I'd understand those spawned in as antags being cleared of negative traits if not negative+positive so there are no balance concerns/no-one aligning their traits for an antag roll to just be disappointed. But those converted should have special clauses rather than removal. I'd want to see a rev have to figure out with their team how to help fight their greatest fear, authority figures.

CosmicScientist commented:
To add onto what coiax said, could there be a fast metabolism/bottomless gullet that speeds up hunger?

coiax commented:
Also, just generally 👍 

Good job Xhuis.

ExcessiveUseOfCobblestone commented:
If you're an antag you should have the traits or else it will be obvious you're antag if you staticname.

Spaceinaba commented:
OH can we have a neutral trait that gives you a surplus replacement limb

Supermichael777 commented:
Should have IC notes in round-start medical records, those need a purpose outside detective.

CosmicScientist commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -457,3 +464,17 @@
+	user.forceMove(src)
+	user.resting = TRUE //so people can jump into crates without slamming the lid on their head
+	close()
+	user.resting = FALSE
```
I have a feeling you should check if it's a crate before resting the user otherwise even though this is consistent behaviour, it isn't consistent with hiding in a locker.

Is it intended balance to not allow someone to hold a stunbaton or revolver in their hand as they do this in a panic?

CosmicScientist replied:
Am I just ignorant and don't know how this works out in practice?

ChangelingRain replied:
this won't cause people to drop items, since it sets the var directly

Xhuis replied:
Yup, it doesn't cause any item dropping.

MMMiracles replied:
You should probably check if people are dragging the closet/crate, as I can totally see people just dragging them around and diving into them once people start shooting. Maybe add a slight stun so jumping in to get out-of-sight of a person is still very viable but jumping in to avoid gunfire is probably a lot less effective since they can just open the locker and shoot you while you're stunned.

RemieRichards commented on code/_onclick/hud/human.dm:
```diff
@@ -87,7 +87,10 @@
 /datum/hud/human/New(mob/living/carbon/human/owner, ui_style = 'icons/mob/screen_midnight.dmi')
 	..()
-	owner.overlay_fullscreen("see_through_darkness", /obj/screen/fullscreen/see_through_darkness)
+	if("Night Vision" in owner.client.prefs.positive_traits) //check prefs directly due to the order in which things are made
```
🤢

Xhuis replied:
HUDs are created before traits are assigned, and I'm not sure of a better way to do this. If you have any advice, feel free to help out.

RemieRichards replied:
consider the possibility that traits may be lost midround one day, as part of a random event or a surgery or whatever.

you're checking whether they should have this trait based on if they WANT it, not if they have it, that's the secondary puke-y thing about this code.

the first one is that it's a string in list check, very bad. Either, change the feature up slightly so that it instead increases the human's eye organ's ```see_in_dark``` (just a ++ will do) or make this nightvision/see_through_dark object be a var on the person's eyes, and make receiving a set of eyes regenerate this part of your hud (shouldn't be difficult)

optimumtact replied:
I thought traits were datums, why are you doing string lists?

Xhuis replied:
The string list is used to track names in prefs with the ``SStraits` lists. When I wrote the code it was faster and cleaner.

RemieRichards commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -653,24 +653,33 @@
-					status = "OK"
-				to_chat(src, "\t <span class='[status == "OK" ? "notice" : "warning"]'>Your [LB.name] is [status].</span>")
+				if(has_trait(TRAIT_SELF_AWARE))
+					status = "[brutedamage] brute damage and [burndamage] fire damage"
```
burn is also cold damage so 'fire' is wrong.

Qbopper commented:
Prosopagnosia (+1): You can't recognize faces, and examining anyone will display their name as Unknown regardless of what they're wearing.

yo holy SHIT I can roleplay as ace

RemieRichards commented on code/modules/mob/living/taste.dm:
```diff
@@ -9,13 +9,15 @@
 /mob/living/carbon/get_taste_sensitivity()
 	var/obj/item/organ/tongue/tongue = getorganslot(ORGAN_SLOT_TONGUE)
-	if(istype(tongue))
+	if(istype(tongue) || has_trait(TRAIT_AGEUSIA))
```
nice nullpointer error.

coiax replied:
Don't you mean !has_trait

davethwave commented:
Passive-minor metabolism boost
Speeds up metabolic rate making chems usage speed up by a third while also also making one hunger faster.

davethwave commented:
Passive-Fat
You are now fat. Though you may suffer less from hunger you are slower and can't fit in a hardsuit. Also gives a another minor trait that gives them a mint phobia due to gibbing effect.

davethwave commented:
Random phobias

XDTM commented:
The skittish effect should definitely be a phobia fear reaction

BeeSting12 commented:
Really fucking fat

You get a fat sprite and you cant wear clothes

davethwave commented:
Positive-man of many languages
Allows you to learn a new language though some you can only understand but not speak

coiax commented:
@XDTM do you think we should make purchasing an imaginary friend or split personality possible with this?

davethwave commented:
Positive- start with maint access
Requires 2-3 points to get. Also doesn't get added if you already had that.

coiax commented:
@davethwave @Xhuis just make learning a "regular" language like Draconic, or EAL cost 1 point.

Spaceinaba commented:
you should be able to understand draconic but when you speak it you speak in broken sentences

XDTM commented:
@coiax definitely not positive traumas, and split personality would likely not work roundstart. Others like concussions don't really make sense, so we'd have to pick the available traumas manually

2Dollarbill commented:
Not a fan of many of these perks. A few of them I'd list out on why I don't like em follows as so: Freerunning, Light Step, Night Vision, self-aware, skittish, just about everything in the negativity side except for nearsighted, and the other neutral stuff is fine.

The way goon handled their traits were very much as they were described, minor or cosmetic additions to your characters. The traits you've picked out can be abused to hell and back. These aren't minor in the slightest if Night Vision gives you an inherent advantage over another player. Some of these are extremely overpriced for the bad downsides they give you. Heavy Sleeper and Prosopagnosia are essentially two free points since the downsides are negligible, however the trade off for the others are so bad, I can't see anyone using it in it's current state. I'd reconsider changing these to be much more cosmetic, like self-aware just makes your character do more "me" commands by itself (example: John Weed breaths in. John Weed breaths out.)

Anonmare commented:
Night vision doesn't actually expand your vision range, it just makes what's already there a bit brighter

Bawhoppen commented:
I just feel like this is the wrong implementation of the system. 
I expected like a long sprawling list of minor stuff like if you are Italian, not some system of balance and point tradeoffs.

Xhuis commented:
It's not the *wrong* implementation of it, just a *different* one. Minor and cosmetic additions can and would work perfectly fine with this system, with no drawbacks to using it in that form. Personally, however, I wanted to make minor gameplay changes to vary the flow of gameplay and make people satisfied by their traits kicking in to make small things different.

Negative traits are inherently more impactful, since I want them to really be *felt.* I'll probably be either making the comparatively low-impact ones more impactful, or making many of the weaker ones less damaging in a typical round. I do want some to be massive tradeoffs on purpose, for people who want to vary their gameplay in unique ways, but I don't want *every* trait to be Pacifist-level detrimental.

Bawhoppen commented:
I am aware it is functionally a correct implementation, but by adding these balance-related changes as the first ever traits, it seems you're setting a precedent on what kind of thing it *should* be for. 
Atleast it feels that way to me.

Xhuis commented:
I don't mind that too much. If it ends up being an issue, it can certainly be toned down, and I'll be keeping an eye out to make sure that no "meta" setup develops since that's not my goal.

Spaceinaba commented:
still needs a robo arm trait

2Dollarbill commented:
Well it's going to go that way with how these perks are currently balanced. They're not minor editions if they can drastically effect how people can do things.

Xhuis commented:
If any of the positive traits cause a major impact to playstyle. please let me know your thoughts and I'll take them to heart. I don't mind negatives as much, because they don't provide inherent buffs.

2Dollarbill commented:
Negatives can be just as much of a bad thing as the upside positives. Having weak positive traits but having comparably bad negative traits is a bad trade off system. Like I said, it'd be better for having these things be cosmetic than affect gameplay. The only thing, imo, that should affect gameplay before the round begins, is the job you selected.

Spaceinaba commented:
self aware and skittish are the worst offenders imo

Bawhoppen commented:
In any case, as other people have mentioned other minor traits we need:
All negative traumas/phobias or however they work
All negative genetic traits (especially stuttering)
Missing limbs/prosthetic limbs
Addictions
Being fat (course we still don't have fat sprites in 2018, but it's acceptable to add)
THE PINEAPPLE PREFERENCE

Xhuis commented:
Self-Aware is significant at a glance, but most people with access to the supplies to heal you have a medical analyzer anyways. I still think it's quite useful, which is why I made it require 2 points, but it isn't meta-defining. Skittish might have have the potential to be a little crazy, admittedly, so I'll be watching especially that to see if there's any changes that need to be made.

Bawhoppen commented:
Self aware doesn't seem that useful, since you don't need to ever really know the exact numbers, and you already are privy to info like "moderate bruising" and such

Anonmare commented:
I think you could weaken skittish by making it stun you for a moment.

Since diving headfirst into a locker is probably gonna knock you a little silly

Bawhoppen commented:
Possible positive trait that seems very useless: Craft at 1.2x speed

Xhuis commented:
Haven't you ever seen an action movie? The only thing it's missing is dodge rolls and slamming down tables for cover.

Xhuis commented on code/datums/traits/negative.dm:
```diff
@@ -0,0 +1,110 @@
+	name = "Nearsighted"
+	desc = "You are nearsighted without prescription glasses, but spawn with a pair."
+	value = -1
+	mob_trait = TRAIT_NEARSIGHT
```
It's just the normal Nearsight trait, just applied so that it prevents being cleansed.

epochayur commented:
not huge on freerunning or lightstep, but this is cool otherwise

coiax commented:
@Xhuis Does Ageusia alter the taste messages that people receive?

Xhuis commented:
Ageusia prevents them from tasting anything at all. They don't get any taste messages, with their only indication of taste being when they eat toxic food and feel sick as a result.

Xhuis commented:
Does anyone know if it's possible to store lists in player saves? For now you have to configure your traits every time at roundstart, but when I tried dabbling with it I couldn't get it to work properly.

Spaceinaba commented:
damn I wish I could pick the limb so I can have a metal left arm like my lord and savior venom snake from the metal gear solid series, specifically the game titled metal gear solid V

Qustinnus commented:
when this is merged moodlets are being PRed and a large group of traits will be added in combo.

Qustinnus commented:
also xhuis ill suck you off for this PR no homo

Kraseo commented:
@2Dollarbill 
![imagen](https://user-images.githubusercontent.com/3855300/35998465-fccef134-0d1c-11e8-98f8-a15c36855c3f.png)

Xhuis commented:
Did some minor menu tweaks; trait names are only bolded when you have them, and they're now stored across rounds so you don't have to redo your setup every time. Once I have medical record integration in, this should be ready to go.

Qustinnus commented:
@Xhuis i think some trait values should be changed, especially the positive traits. theyre a bit cheap which allows for stupid combos which people are using as strawmen as to "why traits wont work"

praisenarsie commented:
![2018-02-08_21-43-21](https://user-images.githubusercontent.com/31222036/35999914-439f10ac-0d19-11e8-8bd0-6063307a9356.png)
make it look like the goon's trait menu or bust

Qustinnus commented:
have you looked at their trait code its a fucking mess

WJohn commented:
How does poor aim work, exactly? I see that it adds 25 degrees (which is a fuckload, guns are basically unusable for you). Does it outright *disable* dual weilding, or just add that 25 degrees on top of the accuracy penalty that dual wielding *already* has, to the point where it would be worthless?

I feel it should probably be dropped to 20 degrees so it's still actually possible to hit somebody in close range with a gun.

Xhuis commented:
I believe Poor Aim skews it by 25 pixels, but I might be wrong. It's approximately the same as dual wielding, which is inaccurate but not impossible to hit with. Point-blank and close range are still accurate the majority of the time.

ike709 commented:
Honestly as the number of traits increases it might be worth breaking up positive/neutral/negative into their own pages.

drawsstuff commented:
Goon has a trait that permanently slurs your speech. I mean, really bad. Something like "This is your captain speaking" becomes "dish isch your capenish schpeashing". I always loved that.
Also, I'd love to see the other phobias become available for selecting. Some people might not pick these disabilities because they want buffs, rather because they just want to have them.

Xhuis commented:
Right now, `/datum/trait` is separate and different from the actual, current "trait" system. The current system uses a list of strings and sources in order to provide those traits in a lightweight way. Trait *datums* (what I'm introducing here) are used to hold effects for roundstart ones, and to apply those string traits where necessary.

Would maintainers prefer some sort of refactoring to migrate the string system over to the new datums, or is this hybrid system that uses datums only when necessary a fine solution?

XDTM commented:
Traits are a replacement for the old disabilities, and their function is to be temporary pseudoflags with multiple source support. The name being the same as the roundstart trait system is just a coincidence, and while it might work for some of those it's not what it's made for.

nicbn commented:
Add a musician trait that is needed to play instruments please

WJohn commented:
How does improved nightvision look in game?

ChuckTheSheep commented:
Having played games with this system, or something similar. Why do the numbers have to be numerically negative for positive traits?

Also, it might not be an issue now, but later on it helps if the acceptable range is "less than absolute 1"
 (-1 to +1). It gives some leeway and allows for more combinations, in the games I've played it's allowed to be totally negative but its really annoying to get 0, and at best you land on -1; so some breathing room up to +1 would help.

Bawhoppen commented:
I think you should actually be allowed to pick as many negative traits as you want. People should be allowed to RP in that manner if they want to.

Xhuis commented:
Picking tons of negative traits would make the game too memey, and also not make much sense to me. You can pick all negative traits for your six, though.

@ChuckTheSheep I don't really follow, what do you mean?

ChuckTheSheep commented:
The first point is just when explaining the system its easier when the positive traits are positive, maybe it's a preference thing.

As for the second, It was a problem the other game had where you rarely get net 0 and often times have to take extra negative traits to get 1 point unspent. Having the ability to get away with 1 point owed might help with variety too?

Iamgoofball commented:
@xhuis let them take all negatives but dont let them get bonus points for it, it'll be fine

granpawalton commented:
Im less concerned with the gameplay effects of this and more concerned with the amount of higher roleplay players this could intrest into the server, these types of players can be very difficult to work with with for some people as they tend to not care about efficient game mechanics as much. Or atleast i dont want this bringing fucking normies into my games who pick a bad trait that costs us a team antag round.

Spaceinaba commented:
the only team antag that traits would massively effect though is nukies though

granpawalton commented:
What if some sec officer takes poor aim or prosoponasia because “muh roleplay” and endangers another officer or himself while going after an antag together. Also what if someone cant call out an antag because social anxiety prevents them from doing it and it means “oh darn! Im a cult now and have valids!” Or “sorry! I couldnt warn you in time because i made a bad decision that i will do again 100%”, which is nothing compared to the almost beneficial one of being able to heal yourself with a welding tool

granpawalton commented:
I mean it could work fine in servers that already have this in effect but reading the traits it makes me fear that the bar for the lowest expectation of another player could be brought down even further

Kraseo commented:
are you genuinely bashing people who pick stuff they want when they have almost no control over the antag pick system

granpawalton commented:
Im genuinely bashing people who dont take responsibility for their actions, and there is a way to control the antag pick system, turn it off for team antag if you are about to do some dumb stuff that could ruin other people’s experience.

XDTM commented:
God forbid some people don't make validhunting their absolute priority

granpawalton commented:
I mean if you want to word it that way, what if are at the end of a round on clock cultist and your sitting in the back not contributing when someone risked their life to convert you, it is kind of unfair to someone who put in the effort to get you but thats what i mean by the lowest expectation you could have of a player which is entirely acceptable

XDTM commented:
If you go out of your way to convert a cripple you get a cripple, and it's no different than converting someone who's new or unrobust; it's actually more evident.

That said, while antag vs crew conflict drives the game, it should not be the only focus and not everything should cater to that. Roleplaying is part of the fun and makes for more varied situations. A cultist who is psychotically terrified of his own cult is not _efficient_, it won't increase your team's chance of winning, but personally i'd find that hilarious, regardless of the team i was on.

F-OS commented on code/datums/traits/negative.dm:
```diff
@@ -0,0 +1,180 @@
+			sleep(27)
+			trait_holder.playsound_local(trait_holder, 'sound/effects/explosion_distant.ogg', 50, FALSE, pressure_affected = FALSE)
+
+/datum/trait/insanity/proc/disclaimer() //I don't /think/ we'll need this but for newbies who think "roleplay as insane" = "license to kill" it's probably a good thing to have
```
Put it in the description, as i learned from my PR ingraining server policy in code is dumb

granpawalton commented:
Im probably just blowing this out of proportion in my mind due to my personal experience with roleplayers. It has worked fine on stations like goon so it might not cause serious issues in tg

kevinz000 commented:
as the chief validhunting powergamer of /tg/station i say you are incorrect @granpawalton and you need to try to rp for once in your powerful valid life.

coiax commented:
> chief validhunting powergamer

I thought that was @Robustin

Kraseo commented:
stinky roleplayers won't let me win the game UGHHH >:(((

Militaires commented:
by the way, dont restrict availability of traits based on antag status or otherwise, the game absolutely should have room for nonsensical unfair things like magic-fearing cultists and paraplegic security teams

dont ruin the american dream with balance

Xhuis commented:
The only restriction right now is Pacifist, which cleanses if you roll antag, for obvious reasons.

MMMiracles commented:
Pacifism would probably only work with regular traitors if they get really creative, otherwise pretty much every other antag roll would be fucked.

XDTM commented:
If you make it a trauma it would be curable through surgery

Qustinnus commented:
traits shouldn't be curable by surgery though, that makes it too easy to get free points.

Xhuis commented:
Unless I messed up somewhere in the code, they aren't. Nearsightedness isn't possible to fix, for instance, iirc.

Qustinnus commented:
Was ment as a reply to XDTM :P

XDTM commented:
While it's true it's also kinda iffy to have non-interactable medical conditions that work exactly like existing interactable ones.

For traumas specifically i'm adding a resilience system in #35520 that allows you to set how hard they are to cure; for example i'd set the roundstart traumas to require lobotomy. It can't be cured right away, and even when cured it would carry the risk of a worse, non-healable brain trauma, but at least you can do _something_ about it.

Qustinnus commented:
Traits aren't traumas and are personality traits. Making them traumas turns this into a game of "Ill just get surgery for free points"

XDTM commented:
Lobotomy causes other, incurable, traumas when performed so it's definitely not free

coiax commented on code/datums/traits/negative.dm:
```diff
@@ -0,0 +1,180 @@
+/datum/trait/nonviolent/on_process()
+	if(trait_holder.mind && trait_holder.mind.antag_datums.len)
+		to_chat(trait_holder, "<span class='boldannounce'>Your antagonistic nature has caused you to renounce your pacifism.</span>")
+		qdel(src)
```
Given that people are concerned that this might be an "antag test", perhaps you could give antagonists an action button to disable it?

Kraseo replied:
how are they actually supposed to "antag test" a pacifist

XDTM replied:
or a pref? "Ignore traits when antagonist"

Xhuis replied:
It's not possible to "antag test" a pacifist because the player could just refuse to attack. Keeping traits as antagonist, though, is both interesting gameplay and switches it up more, so I don't see why not to keep most traits.

coiax replied:
I mean, I suppose they COULD refuse to attack. But it's not like we have the 6th amendment in Space Law (or anyone actually cares about spacelaw). Anyone refusing to punch someone with a tome or punch to test for non-violence, just assume they're guilty and throw them in the permabrig.

Just kidding, no one permas people, just execute them for valids.

coiax commented on code/datums/traits/negative.dm:
```diff
@@ -0,0 +1,180 @@
+/datum/trait/prosthetic_limb
+	name = "Prosthetic Limb"
+	desc = "An accident caused you to lose one of your limbs. Because of this, you now have a random prosthetic!"
+	value = -1
```
I mean, bonus points if you can pick WHICH limb you're missing.

Xhuis replied:
It's doable, but for now this is fine to me.

coiax commented on code/datums/traits/neutral.dm:
```diff
@@ -0,0 +1,27 @@
+	var/datum/species/species = H.dna.species
+	var/liked = species.liked_food
+	species.liked_food = species.disliked_food
+	species.disliked_food = liked
```
Isn't this missing a remove() proc?

Xhuis replied:
Where would I put the remove proc?

coiax replied:
On the /datum/trait/deviant_tastes/remove() ?

Xhuis replied:
Oh, right. I never intend for trait datums to be removed to begin with, but I do suppose it'd be good to be there in case they were.

coiax replied:
I mean for testing and admins messing around in the thunderdome.

coiax commented on code/modules/food_and_drinks/drinks/drinks.dm:
```diff
@@ -394,7 +396,8 @@
 	spillable = FALSE
 	isGlass = FALSE
-	
+	grind_results = list("aluminum" = 10)
```
Did this slip in by accident?

Xhuis replied:
It did, it would seem.

coiax commented on code/modules/food_and_drinks/food.dm:
```diff
@@ -19,13 +19,18 @@
+			if(!H.has_trait(TRAIT_AGEUSIA))
+				if(foodtype & H.dna.species.toxic_food)
+					to_chat(H,"<span class='warning'>What the hell was that thing?!</span>")
+					H.adjust_disgust(25 + 30 * fraction)
```
I think the "food like/dislike" system mixes both preferences and intolerance a lot.

Moths should not be able to digest meat effectively etc.

Qustinnus replied:
If you want I can refactor food likes/dislikes, when I originally made it I was new to byond code

coiax replied:
I'm thinking of giving it a go.

Commit:
Conflicts

Commit:
Unspaghettifies night vision

Xhuis commented:
Some people were asking for reference, so here you go.

### With Night Vision:
![](https://i.imgur.com/S6StpIe.png)

### Without Night Vision:
![](https://i.imgur.com/TImshmp.png)

Anonmare commented:
At first I didn't notice a difference but the one with night vision has one tile more of radius to vision

Xhuis commented:
More accurately, it's *very* slightly brighter, and removes the obscuring "vignette" around your character, which lets you see the tiles diagonally more clearly.

Robustin commented:
You did a reasonable job with this but this but it's just not going to hash out in the long run. 

The only way this will work is by having a strict rule against anything combat-related. Already I see myself taking heavy sleeper and grabbing skittish and self-aware (being able to instantly throw yourself into a locked-locker is quite powerful) and basically getting a new combat ability and improved healing at no real downside. 

Even then this is going to be hard to maintain and I don't really see the upside.

coiax commented:
@Robustin then we make negative traits give fewer points, and positive traits cost more points, based on the BALANCE of the trait.

Also, your example doesn't work, Heavy Sleeper only gives 1 point, and the traits you listed both cost 2 points. Try again.

XDTM commented on code/modules/flufftext/Hallucination.dm:
```diff
@@ -1197,3 +1197,41 @@ GLOBAL_LIST_INIT(hallucinations_major, list(
+/datum/hallucination/rds/supermatter/New(mob/living/carbon/C, forced = TRUE)
+	..()
+	SEND_SOUND(C, 'sound/magic/charge.ogg')
+	to_chat(C, "<span class='boldannounce'>You feel reality distort for a moment...</span>")
```
you could roll this hallucination inside /datum/hallucination/sounds, it already has the malf ai, and nuke op warning, as well as ratvar rising (a bit outated)

XDTM commented on code/modules/flufftext/Hallucination.dm:
```diff
@@ -1197,3 +1197,41 @@ GLOBAL_LIST_INIT(hallucinations_major, list(
+		enter_link = "<a href=?src=[REF(src)];orbit=1>(Click to orbit and freak them out)</a>",
+		source = C, action = NOTIFY_ORBIT)
+	sleep(50)
+	C.see_invisible = INVISIBILITY_OBSERVER
```
this will also show the camera icons, such as the AI, blob, or advanced cameras

coiax replied:
I didn't think that the AI or advanced cameras camera mobs actually had visible icons...

XDTM replied:
scratch that, i think that's handled by an alternate_icon system, so no worries here

AnturK replied:
Still awful.

XDTM commented on code/modules/flufftext/Hallucination.dm:
```diff
@@ -1197,3 +1197,41 @@ GLOBAL_LIST_INIT(hallucinations_major, list(
+	C.playsound_local(C, 'sound/machines/clockcult/ark_deathrattle.ogg', 50, FALSE, pressure_affected = FALSE)
+	C.playsound_local(C, 'sound/effects/clockcult_gateway_disrupted.ogg', 50, FALSE, pressure_affected = FALSE)
+	sleep(27)
+	C.playsound_local(C, 'sound/effects/explosion_distant.ogg', 50, FALSE, pressure_affected = FALSE)
```
same as above

XDTM commented on code/modules/flufftext/Hallucination.dm:
```diff
@@ -1197,3 +1197,41 @@ GLOBAL_LIST_INIT(hallucinations_major, list(
+
+/datum/hallucination/rds/fourth_wall/New(mob/living/carbon/C, forced = TRUE)
+	..()
+	to_chat(C, "<span class='userdanger extremelybig'>[pick("Leave the server" , "Close the game window")] [pick("immediately", "right now")].</span>")
```
fits in /datum/hallucination/message, if you feel like moving it there

Xhuis replied:
`rds` types are easier to manage

AnturK commented on code/controllers/subsystem/processing/traits.dm:
```diff
@@ -0,0 +1,23 @@
+//Used to process and handle roundstart trait datums
+//Due to their nature, roundstart traits are objects, but are otherwise tied to the normal trait system
```
Please don't leave misleading comments, object has very specific meaning.

AnturK commented on code/datums/traits/_trait.dm:
```diff
@@ -0,0 +1,100 @@
+//every trait in this folder should be coded around being applied on spawn
+//these are NOT "mob traits" like GOTTAGOFAST, but exist as a medium to apply them and other different effects
```
So what's a difference between a trait and status effect, because it seems to me you're duplicating the functionality if the primary feature is just processing.

Xhuis replied:
Mainly the tiein with mob traits. It might be possible to make them into a status effect subtype, but it'd require significant changes like forcing it to not go into the mob's status effects.

AnturK replied:
Alright

AnturK commented on code/datums/traits/good.dm:
```diff
@@ -0,0 +1,83 @@
+	lose_text = "<span class='danger'>Everything seems a little darker.</span>"
+
+/datum/trait/night_vision/on_spawn()
+	var/mob/living/carbon/human/H = trait_holder
```
So is it livings or humans, typecheck

Xhuis replied:
Traits have a human requirement by default in their base type that ensures that on_spawn won't be called if it's not a human, since the trait won't be assigned.

coiax replied:
But I see some of these traits could be assigned perfectly well to livings, rather than just humans.

Xhuis replied:
They're coded around human requirements due to how the code handles them.

Now that I consider it, the system could likely stand to use significant improvement...

AnturK commented on code/modules/client/preferences.dm:
```diff
@@ -761,6 +771,68 @@ GLOBAL_LIST_EMPTY(preferences_datums)
+	if(!SStraits)
+		to_chat(user, "<span class='danger'>The trait subsystem is still initializing! Try again in a minute.</span>")
+		return
+	if(SSticker.HasRoundStarted() && !isnewplayer(user))
```
Why. This is inconsistent with any other preferences.

Xhuis replied:
If someone is cloned during a round I suspect that it'd change their traits midround based on what they changed between dying and being cloned, due to how body transference works.

coiax replied:
Just cache whatever they have at the start of the round.

AnturK commented on code/modules/client/preferences.dm:
```diff
@@ -808,6 +880,61 @@ GLOBAL_LIST_EMPTY(preferences_datums)
+						positive_traits -= trait
+						all_traits -= trait
+					else if(trait in negative_traits)
+						if(balance + value < 0)
```
The whole trait composition should be checked once instead of this giant IF tree per trait change in my opinion.

Xhuis replied:
What?

AnturK replied:
Just do it once when all traits are to be saved instead of every change. Would make it easier to read.

Xhuis replied:
I'm not sure I follow... as in, cache the number and then update it when traits are saved, or don't change the number at all until they're saved?

AnturK replied:
Just have a single validation function that checks the overall state.
```
DoWhatever() //Or after any other operation
Validate() //This will show some appropriate error messages. (Alerts are kinda clunky, better show it on the actual interface which parts are wrong)
..

if(Validate())
 Save()

```

Xhuis replied:
I'm having some issues getting this to work, by way of it not assigning positive traits at all. Here are the procs I'm using:

```
	else if(href_list["preference"] == "trait")
		if(SSticker.HasRoundStarted() && !isnewplayer(user))
			to_chat(user, "<span class='danger'>The round has already started. Please wait until next round to set up your traits!</span>")
			return
		switch(href_list["task"])
			if("close")
				user << browse(null, "window=mob_occupation")
				ShowChoices(user)
			if("update")
				var/trait = href_list["trait"]
				var/value = SStraits.trait_points[trait]
				if(ValidateTraits(trait, value))
					SaveTraits(trait, value)
				SetTraits(user)
```

ValidateTraits:
```
/datum/preferences/proc/ValidateTraits(trait, value) //ensures that you can take or lose a trait, returning FALSE if you can't
	if(isnull(value))
		return
	var/bal = GetTraitBalance()
	if(trait in all_traits && bal + value < 0)
		to_chat(usr, "<span class='warning'>Refunding this would cause your balance to go below zero!</span>")
		return
	else if(bal - value < 0)
		to_chat(usr, "<span class='warning'>You don't have the balance to take this!</span>")
		return
	else if(all_traits.len >= MAX_TRAITS)
		to_chat(usr, "<span class='warning'>You can only have [MAX_TRAITS] total traits!</span>")
		return
	return TRUE
```

SaveTraits:
```
/datum/preferences/proc/SaveTraits(trait, value) //adds the provided trait to the trait list
	var/list/category = list()
	if(value < 0)
		category = negative_traits
		to_chat(world, "Neg cat")
	else if(value == 0)
		category = neutral_traits
		to_chat(world, "Neu cat")
	else if(value > 0)
		category = positive_traits
		to_chat(world, "Pos cat")
	if(!category.len) //invalid value
		to_chat(world, "No cat")
		return
	if(trait in category)
		category -= trait
		all_traits -= trait
		to_chat(world, "Remove trait")
	else
		category += trait
		all_traits += trait
		to_chat(world, "Add trait")
```

The issue seems to be stemming in SaveTraits, as ValidateTraits passes fine. The `to_chat(world, X)` are for debug purposes. Can you see any immediate errors from just skimming through that might cause that?

AnturK commented on code/modules/mob/living/status_procs.dm:
```diff
@@ -168,6 +185,8 @@
 
 /mob/living/proc/has_trait(trait, list/sources)
+	if(roundstart_traits[trait])
+		return roundstart_traits[trait] //fetch the trait object directly in case we need to modify it
```
This really shouldn't have different return types depending if you check one define or another

AnturK reviewed (changes requested):
I'm okay with the idea but the code is unmaintainable mess.

Kevak commented:
Add monophobia for a negative trait. You already have the code from brain traumas. It would fit.

Commit:
Fuck everything

ChuckTheSheep commented on code/modules/client/preferences.dm:
```diff
@@ -1405,6 +1404,9 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 /datum/preferences/proc/copy_to(mob/living/carbon/human/character, icon_updates = 1)
+	if(!generated_char_traits)
+		generated_char_traits = TRUE //usually it would be better to have a check for character_traits.len but that doesn't FUCKING WORK and I don't know why.
+		character_traits = all_traits
```
Isn't len being phased out for length() anyway?

Xhuis replied:
If so, I've not heard of it until now.

ChuckTheSheep replied:
It's a BYOND thing, but it's been a 'thing' for a few years now. I think it says it in the reference guide.
But, does it work with length()?

Xhuis replied:
I checked the reference guide and didn't see any particular pointers about `.len` being inferior to `length()`, and I've used `.len` for a really long time, to the point where I've never actually used `length()` to check a list. I imagine it would work, though.

Doppeltongue commented:
Traits should probably show up in medical records.

Kraseo commented:
they're planned to

Mey-Ha-Zah commented:
>Test merged
>No button to access them

Xhuis commented:
Why would you testmerge this while I'm asleep.

Xhuis commented:
Oh, I think it's probably because it needs a config option checked to use. I don't think testmerges use configs. I can probably make a temporary workaround for testing.

Qustinnus commented:
be smart like me and make the config "turn off traits"

Commit:
Issues patch

Xhuis commented:
Made some changes based on the testmerge on bagil
* used the trait subsystem for assigning and generating lists
* re-disallowed people from editing traits mid-game due to really fucking weird issues with checks always passing even with a binary `if(FALSE)` replacement for testing
* fixed night vision
* made skittish respect the laws of physics
* nearsighted should be incurable now
* removed arachnophobia
* added a disclaimer message in prosthetic limb that tells how to repair it and what limb
* halved the frequency of social anxiety stuttering

overall I'm happy with how things are going so far and player reception for the first few rounds was positive with the exception of the skittish and night vision bugs

Xhuis commented:
This is not stale, I pushed a commit not a day ago. My responses to maintainer review are waiting for a response so I do them right.

optimumtact commented:
too bad so sad, I am the skeleton of slow commits and I am here to rattle your pr

Xhuis commented:
consider me thoroughly spooked

GunHog commented:
I have to agree with that. You choose to limit yourself. If you roll antag, you will have to get creative in order to murderbone! You clearly picked 'pacifist' to have a challenge, right?

Commit:
Merge branch 'master' into traits

davethwave commented:
Honestly you should be able to choose if you want certain traits if you are a traitor or not. Such as a 3 button system I guess, on, on but off if antag, buy.

Commit:
Merge branch 'master' into traits

Qustinnus commented:
pls get this merged this before my PR gets the stale tag :^) oranges is ruthless

Xhuis commented:
oh fuck it's happening

Xhuis commented:
I'll be listening closely for people's input on these. Unbalanced ones or ones that should be tweaked will be adjusted as needed, and if the feature proves to be fun and stable, then people are free to add more!

Qustinnus commented:
AWWWWW YESSSS

MMMiracles commented:
awh yeee boiiiiiiiiiii

Nitrousoxide commented:
I would suggest adding a popup alert if you are selecting a trait that makes you more or less completely unable to do certain jobs and you select that job after a round has started.  For instance, pacifist and miner/sec.

Don't keep them from playing the job, if they want to roleplay a pacifist sec officer go for it.  But give people a heads up before they (potentially) join as someone who basically can't do their job.

AlexTheSysop42 commented:
You can still use stunbatons as sec, mining also isnt completely impossible because of the sonic jackhammer/drills/pickaxes

epochayur commented:
if you pick pacifist and have sec enabled, that's your own fault

Xhuis commented:
I'm fairly certain pacifist miners can actually use resonators, too.

PKPenguin321 commented:
does this have support for conflicting traits? i want to add a trait that makes you not get knocked down by melee weapons as often and another that makes it happen *more* often and it would be weird if you could have both at the same time, so it would be cool if i could restrict the player from having one if he's already taken the other

Xhuis commented:
No support for conflicting traits right now, but it's easily doable.
